tracer/src/Datadog.Trace/ClrProfiler/AutomaticTracer.cs

```diff
@@ -0,0 +1,69 @@
+// <copyright file="AutomaticTracer.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System.Collections.Generic;
+using System.Threading;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.ClrProfiler
+{
+    internal class AutomaticTracer : CommonTracer, IAutomaticTracer, IDistributedTracer
+    {
+        private static readonly AsyncLocal<IReadOnlyDictionary<string, string>> DistributedTrace = new();
```

Comment:
I feel like this may need a `AsyncLocal<Stack<IReadonlyDictionary<string,string>>>` that way we have an AsyncLocal way to determine which `SpanContext` object is the most recent, and then whenever a SpanContext is created on either side, we just push it to the Stack. This could solve the problem where nested manual spans are created, so that instead of checking the result of `GetDistributedTrace` below to see what `SpanContext` should be the parent, they just get the value at the top of this stack and convert it to a local `SpanContext` if it originated from the other side of the Manual/Automatic divide. What do you think?

Reply:
I broke it when trying to optimize it (I guess we need a test for this scenario). A stack isn't needed.

`GetDistributedTrace` in the AutomaticTracer should return `DistributedTrace.Value`, and `SetSpanContext` should update the asynclocal value. `GetSpanContext` (in both automatic and manual) should also have a check to see if the value retrieved is an instance of SpanContext (in which case the context is local so we should just return null without calling `SpanContextPropagator.Instance.Extract`).

That's what I did originally, but then I tried to remove the asynclocal write in AutomaticTracer when there is no ManualTracer (= `_child` is null), and I went too far :( Please make sure that `SetDistributedTrace` remains a no-op in `AutomaticTracer` when `_child` is null

Reply:
Changes made in 32f782bd6245d63eda46491823c0fdba498fb949 . This should work and this is tested with a regression test introduced in 90caad049e81d11dbffd04bb06b73b12013be2e7